### PR TITLE
[improvement] Add concurrency limits and load balancing

### PR DESCRIPTION
**Pull request by ellisjoe**
## After this PR
Adds a ConcurrencyLimitingChannel that only allows calls to be executed if there is sufficient capacity available on the underlying channel. Also adds a RoundRobinChannel which rotates through channels when sending requests while skipping over channels that don't have any available capacity.

`LimitedChannel` may be of some interest.

Think we'll want to add another thing around the RoundRobinChannel which has some fallback behavior (eg: send the request to a random host when non of the LimitedChannels have any capacity, or queue it, etc.)

### CODE REVIEW

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                .build();
+        Limiter<Object> limiter = SimpleLimiter.newBuilder()
+                .limit(WindowedLimit.newBuilder().build(aimdLimit))
+                .build();
+        return new ConcurrencyLimitedChannel(delegate, limiter);
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeCreateCall(Endpoint endpoint, Request request) {
+        return limiter.acquire(CONTEXT).map(listener -> {
```
Comment (markelliot) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java]: limiters should be keyed by endpoint

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A channel that monitors the successes and failures of requests in order to determine the number of concurrent
+ * requests allowed to a particular channel. If the channel's concurrency limit has been reach, the call will not be
```
Comment (markelliot) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java]: if the channel's concurrency limit has been reached, the maybeCreateCall method returns empty

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java (hunk 1 above)
Comment (ellisjoe) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java]: done

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java (hunk 2 above)
Comment (ellisjoe) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java]: done

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    public RoundRobinChannel(List<LimitedChannel> delegates) {
+        this.delegates = delegates;
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeCreateCall(Endpoint endpoint, Request request) {
+        int host = currentHost.getAndUpdate(value -> toIndex(value + 1));
+
+        for (int i = 0; i < delegates.size(); i++) {
+            LimitedChannel channel = delegates.get(toIndex(host + i));
```
Comment (uschi2000) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:50]: should take an immutable copy of this list to avoid the race condition

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            if (DROP_CODES.contains(result.code())) {
+                listener.onDropped();
+            } else {
+                listener.onSuccess();
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable throwable) {
+            listener.onIgnore();
```
Comment (uschi2000) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java:110]: what would a channel timeout or a connection refused exception mean for the limiter?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java (hunk 3 above)
Comment (uschi2000) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:50]: (or if you want the list to be mutable you should document that and remove the race)

Comment (uschi2000) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:50]: (to be explicit: you can get a out of bound exception if the host list shrinks under you)

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        when(delegate.createCall(endpoint, request)).thenReturn(responseFuture);
+    }
+
+    @Test
+    public void testLimiterAvailable_successfulRequest() {
+        limitAvailable();
+        responseCode(200);
+
+        assertThat(channel.maybeCreateCall(endpoint, request)).contains(responseFuture);
+
```
Comment (uschi2000) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java:61]: nit: would remove this line break (same below)

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    @Test
+    public void testUnavailable() {
+        limitUnvailable();
+
+        assertThat(channel.maybeCreateCall(endpoint, request)).isEmpty();
+
+        verifyZeroInteractions(listener);
+    }
+
+    private void responseCode(int code) {
```
Comment (uschi2000) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java]: nit: would call this mockResponseCode (same below)

File: dialogue-core/src/main/java/com/palantir/dialogue/core/LimitedChannel.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import com.palantir.dialogue.Response;
+import java.util.Optional;
+
+/**
+ * Similar to {@link com.palantir.dialogue.Channel}, but may not actually execute the call (eg: when the channel is
+ * overloaded).
+ */
+public interface LimitedChannel {
+    Optional<ListenableFuture<Response>> maybeCreateCall(Endpoint endpoint, Request request);
+}
```
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/LimitedChannel.java:31]: I'm not sure this should be public since we allow pre-flight request limiting to be disabled entirely.

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Round robins requests across many channels, attempting to choose a channel that has some available capacity.
+ */
+public final class RoundRobinChannel implements LimitedChannel {
+
+    private final AtomicInteger currentHost = new AtomicInteger(0);
+    private final List<LimitedChannel> delegates;
+
+    public RoundRobinChannel(List<LimitedChannel> delegates) {
```
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:37]: I think we're missing a concept of a `ChannelSelector`. We will need to force rotation when requests fail in certain ways, but this is only aware of pre-flight limiting, and doesn't seem like the right place for exceptional response interpretation.

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Round robins requests across many channels, attempting to choose a channel that has some available capacity.
+ */
+public final class RoundRobinChannel implements LimitedChannel {
```
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:32]: Should this implement channel?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java (hunk 4 above)
Comment (markelliot) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java:110]: this matches today's behavior, which is that timeouts + connection refused do not actually adjust the limiter

File: dialogue-core/src/main/java/com/palantir/dialogue/core/LimitedChannel.java (hunk 7 above)
Comment (markelliot) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/LimitedChannel.java:31]: I think the idea here is that you'd wire the thing differently rather than configure the channels to behave differently

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java (hunk 9 above)
Comment (ellisjoe) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:32]: I'd like to have something wrap this and turn it into a channel, and in there you can do things like queue requests when there isn't enough space on the underlying channel to execute the request at this moment.

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java (hunk 8 above)
Comment (ellisjoe) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:37]: Think you just end up with another implementation of the `LimitedChannel` which returns Optional.empty() for some period of time after a particular exception. The goal here is to back each of these things into very separate layers to avoid entangling things like they are today in remoting

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java (hunk 3 above)
Comment (ellisjoe) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java:50]: 👍 done

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java (hunk 5 above)
Comment (ellisjoe) [reply dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java:61]: done

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java (hunk 6 above)
Comment (ellisjoe) [reply dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java]: done

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+                // Don't count slow calls as a sign of the server being overloaded
+                .timeout(Long.MAX_VALUE, TimeUnit.DAYS)
+                .build();
+        return SimpleLimiter.newBuilder()
+                .limit(WindowedLimit.newBuilder().build(aimdLimit))
+                .build();
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeCreateCall(Endpoint endpoint, Request request) {
```
Comment (uschi2000) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java]: maybeExecutor (as per other PR).

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * A channel that monitors the successes and failures of requests in order to determine the number of concurrent
+ * requests allowed to a particular channel. If the channel's concurrency limit has been reached, the maybeCreateCall
```
Comment (uschi2000) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java]: `{@link #maybeExecute}`

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+    private final AtomicInteger currentHost = new AtomicInteger(0);
+    private final List<LimitedChannel> delegates;
```
Comment (uschi2000) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java]: nit: could make this an ImmutableList to make the contract clear

File: dialogue-core/src/test/java/com/palantir/dialogue/core/RoundRobinChannelTest.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    @Test
+    public void testIgnoresUnavailableChannels() {
+        when(channelA.maybeCreateCall(endpoint, request)).thenReturn(UNAVAILABLE);
+
+        assertThat(loadBalancer.maybeCreateCall(endpoint, request)).isEqualTo(CHANNEL_B_RESPONSE);
+        assertThat(loadBalancer.maybeCreateCall(endpoint, request)).isEqualTo(CHANNEL_B_RESPONSE);
+    }
+
+    @Test
+    public void testNoChannelsAvailable() {
```
Comment (uschi2000) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/RoundRobinChannelTest.java:72]: nit: could also test the empty case `new RoundRobinChannel(ImmutableList.of())`

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        verify(listener).onIgnore();
+    }
+
+    @Test
+    public void testUnavailable() {
+        mockLimitUnavailable();
+
+        assertThat(channel.maybeCreateCall(endpoint, request)).isEmpty();
+        verifyZeroInteractions(listener);
+    }
```
Comment (uschi2000) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java:99]: let's also add one test that uses a standard ConcurrencyLimitingChannel#create channel to test the default behavior.

**Review by uschi2000 — APPROVED** [review]
let's run with this and see where we end up.

File: dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java (hunk 12 above)
Comment (ellisjoe) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/RoundRobinChannel.java]: done

File: dialogue-core/src/test/java/com/palantir/dialogue/core/RoundRobinChannelTest.java (hunk 13 above)
Comment (ellisjoe) [reply dialogue-core/src/test/java/com/palantir/dialogue/core/RoundRobinChannelTest.java:72]: good call. done

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java (hunk 14 above)
Comment (ellisjoe) [reply dialogue-core/src/test/java/com/palantir/dialogue/core/ConcurrencyLimitedChannelTest.java:99]: done
